feat: Dynamo LLM Router Integeration

arunraman opened the pull request:
This pull request introduces a new example deployment guide and configuration for integrating NVIDIA Dynamo with an intelligent LLM Router. The changes add platform-specific deployment manifests, Helm values, and routing configuration to enable production-ready, multi-model LLM request routing using Dynamo. The deployment patterns support intelligent routing across multiple models based on task type and complexity.

**LLM Router integration and deployment:**

* Added a new section to the `examples/README.md` linking to the LLM Router deployment guide for NVIDIA Dynamo integration, making it discoverable for users seeking intelligent routing solutions.

**Kubernetes deployment manifests:**

* Introduced three new Kubernetes YAML manifests in `examples/deployments/LLM Router/`:
  * [`agg.yaml`](diffhunk://#diff-4ac1c310a4206ca370632cc12b11655bd658b2b86c86e2e0f9d118801eda2de3R1-R26): Defines a single `VllmDecodeWorker` service for aggregated model serving.
  * [`disagg.yaml`](diffhunk://#diff-b5930f4b5a5cabe21632c4c7357f748fba8f3fa752f8df8cb7e39c206900b23bR1-R43): Deploys both `VllmDecodeWorker` and `VllmPrefillWorker` services for disaggregated model serving, allowing for more granular scaling.
  * [`frontend.yaml`](diffhunk://#diff-0fa6c7c831cb3cfcc4b793969b92baf537bde58889aaea91469c056505f2bc50R1-R16): Specifies the frontend service for the LLM Router, connecting to the Dynamo backend.

**Helm values and configuration:**

* Added `llm-router-values-override.yaml` to provide Helm chart values for deploying the LLM Router with Dynamo integration, including image, environment, service, and storage configuration, as well as support for external ConfigMap-based routing configuration.

**Intelligent LLM routing policies:**

* Added `router-config-dynamo.yaml` ConfigMap manifest, defining intelligent routing policies that map various task types and complexity levels to specific LLMs (Llama 8B, Llama 70B, Mixtral 8x22B), leveraging environment variables for service endpoints and authentication.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Added end-to-end LLM Router deployment examples for Kubernetes with NVIDIA Dynamo integration.
  - Supports task-based and complexity-based routing through a unified API gateway.
  - Includes shared frontend plus aggregated and disaggregated model worker options.
  - Provides configurable Helm values and routing policies with demo app support.

- Documentation
  - Introduced a new production deployment guide for LLM Router with step-by-step installation, requirements, validation, and troubleshooting.
  - Linked the guide under platform-specific deployment references.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

arunraman committed:
feat: add LLM Router deployment examples and configuration for NVIDIA Dynamo integration

- Updated README.md to include a new section for LLM Router deployment with NVIDIA Dynamo.
- Added new YAML files for aggregated and disaggregated worker configurations (agg.yaml, disagg.yaml).
- Introduced frontend.yaml for shared API frontend service.
- Created router-config-dynamo.yaml for routing policies and model configurations.
- Added llm-router-values-override.yaml for Helm values specific to LLM Router integration.
- Included comprehensive documentation on deployment steps and routing strategies.

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

arunraman committed:
refactor: remove deprecated helm-enhancement-implementation.yaml and update README.md for model routing

- Deleted the helm-enhancement-implementation.yaml file as it is no longer needed.
- Updated README.md to reflect changes in model routing, including new API base URLs and model names.
- Adjusted environment variable descriptions for clarity, particularly regarding the DYNAMO_API_KEY for local deployments.
- Enhanced deployment instructions to include multiple model deployment examples.

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

arunraman committed:
chore: update LLM Router configuration files and README for improved deployment instructions

- Added SPDX license headers to llm-router-values-override.yaml.
- Updated imageRegistry placeholder in llm-router-values-override.yaml for clarity.
- Revised README.md to reflect changes in directory structure and emphasize the need to update imageRegistry and imagePullSecrets.
- Adjusted paths in README.md for configuration file references to ensure accuracy.
- Modified router-config-dynamo.yaml to enhance model routing strategies and updated model names for better clarity.

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

arunraman committed:
Fix trailing whitespace in YAML files for pre-commit checks

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

arunraman committed:
Update README.md to remove trailing whitespace and enhance clarity in deployment instructions

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

mohammedabdulwahhab reviewed (approved):
A few small nits but overall LGTM

mohammedabdulwahhab commented on examples/deployments/LLM Router/README.md:
```diff
@@ -0,0 +1,1223 @@
+# LLM Router with NVIDIA Dynamo Cloud Platform
+## Kubernetes Deployment Guide
+
```
I'd restructure to:

- small overview of router + dynamo
- quickstart
- longer reference/details

mohammedabdulwahhab replied:
Makes for better reading UX but this can be a second pass

arunraman committed:
Update LLM Router README.md for improved clarity and deployment instructions

- Revised the overview and table of contents for better organization.
- Enhanced quickstart section with detailed environment variable setup and deployment steps.
- Updated routing strategies and API usage examples for clarity.
- Adjusted version numbers and image references to reflect the latest updates.
- Removed outdated sections and ensured consistency throughout the document.

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

arunraman committed:
Update LLM Router README.md to fix trailing whitespace in routing strategies section for improved readability

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

arunraman committed:
chore: trigger CI re-run

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

arunraman committed:
fix: correct documentation links to use docs/kubernetes path

Signed-off-by: arunraman <arunraman.19@gmail.com>
Signed-off-by: arunraman <arraman@nvidia.com>

arunraman committed:
refactor: rename LLM Router directory to LLMRouter

- Rename 'examples/deployments/LLM Router' to 'examples/deployments/LLMRouter'
- Remove spaces from directory name for better Linux/Mac compatibility
- Update all references in examples/README.md and deployment files
- Update cd commands to use new path without quotes

Signed-off-by: arunraman <arraman@nvidia.com>

rmccorm4 commented:
Closing due to lack of activity
